src/Symfony/Component/Security/Http/Logout/LogoutUrlGenerator.php

```diff
@@ -128,4 +122,56 @@ private function generateLogoutUrl($key, $referenceType)
 
         return $url;
     }
+
+    /**
+     * @param string|null $key The firewall key or null use the current firewall key
+     *
+     * @return array The logout listener found
+     *
+     * @throws \InvalidArgumentException if no LogoutListener is registered for the key or could not be found automatically.
+     */
+    private function getListener($key)
+    {
+        $guess = $key === null;
+
+        // Fetch the current provider key from token, if possible
+        if ($guess && null !== $this->tokenStorage) {
+            $token = $this->tokenStorage->getToken();
+
+            if ($token instanceof AnonymousToken) {
+                throw new \InvalidArgumentException('Unable to generate a logout url for an anonymous token.');
+            }
+
+            if (null !== $token && method_exists($token, 'getProviderKey')) {
+                $key = $token->getProviderKey();
```

Comment:
Then here `if (isset($this->listeners[$key]) { return $this->listeners[$key]; }`